### PR TITLE
Fix multiple memory corruption errors revealed by fuzzing

### PR DESCRIPTION
**Pull request by sashashura**
Fixes the following crashes:
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=32178
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=32228
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=33217
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=33762
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=34134
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=35640
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=36663
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=37041
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=37125
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=39438
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40058
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40120
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=41152

### CODE REVIEW

**Comment by SunBlack** [comment]
As using `memcpy` is not anymore recommend and you touching a lot of this lines: Can you rebase your MR to this #4299 (or wait until #4299 is merged and rebase it then) as there will be a lot of merge conflicts?

**Comment by sashashura** [comment]
It is not that bad. https://github.com/PointCloudLibrary/pcl/pull/4299 didn't touch any memcpy in ply_io.cpp
There are only 5 changes: one include, two trailing spaces removed and two memsets replaced with fills.
But there you go - the merged version.

**Comment by sashashura** [comment]
Does anything block it from being merged?

**Comment by sashashura** [comment]
Related to https://github.com/PointCloudLibrary/pcl/pull/5333 and https://github.com/PointCloudLibrary/pcl/pull/5332 previously reviewed by @mvieth

**Comment by sashashura** [comment]
If the idea is to keep this pull request until https://github.com/PointCloudLibrary/pcl/pull/4299 is merged, that PR was started in 2020 and is about coding style, meanwhile my PR fixes 13 unique crashes that happen because of the already existing memory corruptions.

**Comment by mvieth** [comment]
Sorry for the delay, there were some other PRs and issues that needed attention.
So I think it could make sense to add functions to PCLPointCloud2 for accessing data with bound checking. This could reduce code duplication and would also make code shorter. My idea would be something like this:
```cpp
    template<typename T> inline
    const T& at(const pcl::uindex_t& point_index, const pcl::uindex_t& field_offset) const {
      const auto position = point_index * point_step + field_offset;
      if (data.size () >= (position + sizeof(T)))
        return reinterpret_cast<const T&>(data[position]);
      else
        throw std::out_of_range("PCLPointCloud2::at");
    }

    template<typename T> inline
    T& at(const pcl::uindex_t& point_index, const pcl::uindex_t& field_offset) {
      const auto position = point_index * point_step + field_offset;
      if (data.size () >= (position + sizeof(T)))
        return reinterpret_cast<T&>(data[position]);
      else
        throw std::out_of_range("PCLPointCloud2::at");
    }
```
These mimic [std::vector::at](https://en.cppreference.com/w/cpp/container/vector/at). The reinterpret_cast replaces the memcpy needed otherwise. Both reading and writing is possible, usage would e.g. be like this: `cloud.at<float>(0, y_offset) = 1.0f;` (setting y coordinate of first point to 1).
@kunaltyagi @larshg @SunBlack What do you think?

**Comment by sashashura** [comment]
Although the question wasn't for me, I think it is a great idea. The functions became shorter, more readable and no non-necessary memcpy. I went ahead and implemented your suggestion.

**Comment by larshg** [comment]
Looks good.
@sashashura there are places where you still have memcpy, but I think you can use the new method to also assign there. Like @mvieth also wrote in the original proposal?

**Comment by larshg** [comment]
Ie.
`memcpy (&cloud_->at<Scalar>(vertex_count_, vertex_offset_before_),
              &value,
              sizeof (Scalar));`

To
`cloud_->at<Scalar>(vertex_count_, vertex_offset_before_) = value;`

**Comment by sashashura** [comment]
Good catch! Missed the opportunity. Fixed.

**Comment by sashashura** [comment]
Resolved merge conflict because of edited whitespace. Can we finally merge it?

**Review by larshg — APPROVED** [review]

File: io/src/ply_io.cpp
Comment (mvieth) [inline io/src/ply_io.cpp:1660]:
```suggestion
      PCL_ERROR ("[pcl::io::savePLYFileBinary] Input point cloud has no XYZ data!\n");
```

File: io/src/ply_io.cpp
Comment (sashashura) [inline io/src/ply_io.cpp:1018]:
```suggestion
              fs << static_cast<int>(color.r) << " " << static_cast<int>(color.g) << " " << static_cast<int>(color.b) << " " << static_cast<int>(color.a);
```

File: io/src/ply_io.cpp
Comment (sashashura) [inline io/src/ply_io.cpp:1153]:
```suggestion
              line << static_cast<int>(color.r) << " " << static_cast<int>(color.g) << " " << static_cast<int>(color.b);
```

File: io/src/ply_io.cpp
Comment (sashashura) [inline io/src/ply_io.cpp:1133]:
```suggestion
              line << static_cast<int>(color.r) << " " << static_cast<int>(color.g) << " " << static_cast<int>(color.b) << " " << static_cast<int>(color.a);
```

File: io/src/ply_io.cpp
Comment (sashashura) [inline io/src/ply_io.cpp:1031]:
```suggestion
              fs << static_cast<int>(color.r) << " " << static_cast<int>(color.g) << " " << static_cast<int>(color.b);
```

File: io/src/ply_io.cpp
Comment (mvieth) [inline io/src/ply_io.cpp:290]:
```suggestion
      PCL_WARN ("[pcl::PLYReader::vertexScalarPropertyCallback] Incorrect data index specified (%lu)!\n", vertex_count_ * cloud_->point_step + vertex_offset_before_);
```

File: io/src/ply_io.cpp
Comment (mvieth) [inline io/src/ply_io.cpp:321]:
```suggestion
      PCL_WARN ("[pcl::PLYReader::vertexListPropertyContentCallback] Incorrect data index specified (%lu)!\n", vertex_count_ * cloud_->point_step + vertex_offset_before_);
```

File: io/src/ply_io.cpp
Comment (mvieth) [inline io/src/ply_io.cpp:393]:
```suggestion
      PCL_WARN ("[pcl::PLYReader::vertexColorCallback] Incorrect data index specified (%lu)!\n", vertex_count_ * cloud_->point_step + rgb_offset_before_);
```

**Review by mvieth — APPROVED** [review]
Thank you for your contribution!

**Review by larshg — APPROVED** [review]

**Pull request merged**
